python/ray/data/_internal/execution/streaming_executor.py

```diff
@@ -0,0 +1,94 @@
+import logging
+from typing import Iterator, Optional
+
+from ray.data._internal.execution.interfaces import (
+    Executor,
+    ExecutionOptions,
+    RefBundle,
+    PhysicalOperator,
+)
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.execution.streaming_executor_state import (
+    Topology,
+    OpState,
+    build_streaming_topology,
+    process_completed_tasks,
+    select_operator_to_run,
+)
+from ray.data._internal.stats import DatasetStats
+
+logger = logging.getLogger(__name__)
+
+
+class StreamingExecutor(Executor):
+    """A streaming Dataset executor.
+
+    This implementation executes Dataset DAGs in a fully streamed way. It runs
+    by setting up the operator topology, and then routing blocks through operators in
+    a way that maximizes throughput under resource constraints.
+    """
+
+    def __init__(self, options: ExecutionOptions):
+        # TODO: implement stats recording. We might want to mutate a single stats
+        # object as data is streamed through (similar to how iterating over the output
+        # data updates the stats object in legacy code).
+        self._stats: Optional[DatasetStats] = None
+        super().__init__(options)
+
+    def execute(
+        self, dag: PhysicalOperator, initial_stats: Optional[DatasetStats] = None
```

Comment:
`initial_stats` is not used, can we have a TODO or just remove it?

Reply:
This will be needed to implement the TODO.